### PR TITLE
QgsFeature python optimisations

### PR DESCRIPTION
**Pull request by nyalldawson**
Optimise some more frequently called methods for use in PyQGIS

### CODE REVIEW

File: python/PyQt6/core/auto_generated/qgsattributes.sip.in [hunk 1]
```diff
@@ -72,14 +72,15 @@ typedef QVector<QVariant> QgsAttributes;
...
+  qv->resize( listSize );
+  QVariant *outData = qv->data();
 
   for ( Py_ssize_t i = 0; i < listSize; ++i )
   {
     PyObject *obj = PyList_GET_ITEM( sipPy, i );
     if ( obj == Py_None )
     {
-      qv->append( QVariant( QVariant::Int ) );
+      *outData++ = QVariant( QVariant::Int );
```
Comment (troopa81) [inline python/PyQt6/core/auto_generated/qgsattributes.sip.in:83]: Are we gaining something relevant here ?

**Comment by ptitjano** [comment]
@nyalldawson Thanks. Have you been able to test it and measure the improvement?

File: python/PyQt6/core/auto_generated/qgsattributes.sip.in (hunk 1 above)
Comment (nyalldawson) [reply python/PyQt6/core/auto_generated/qgsattributes.sip.in:83]: Yes -- it reduces the qt implicitly shared container detachment checks from happening once per attribute appended to just once upfront. Ie both QVector::append and QVector::data check for shared use and detachment **every** time they are called, even if it's in a trivial loop like this. These detachment checks add up quickly!

**Comment by nyalldawson** [comment]
@ptitjano 

> Have you been able to test it and measure the improvement?

With a similar test to the loop from https://github.com/qgis/QGIS/pull/57470, calling 10.5 million times

Before: 
18.3s

After:
16.4s

With the optimisations from https://github.com/qgis/QGIS/pull/57470:

Before (#57470 alone):
12.8s

After (this + #57470):
11.7s


So no where near as dramatic as #57470, but still worthwhile!

File: python/PyQt6/core/auto_generated/qgsattributes.sip.in [hunk 2]
```diff
@@ -70,16 +70,21 @@ typedef QVector<QVariant> QgsAttributes;
     return 1;
   }
 
-  QgsAttributes *qv = new QgsAttributes;
   Py_ssize_t listSize = PyList_GET_SIZE( sipPy );
-  qv->reserve( listSize );
+  // Initialize attributes to null. This has two motivations:
+  // 1. It speeds up the QVector construction, as otherwise we are creating n default QVariant objects (default QVariant constructor is not free!)
+  // 2. It lets us shortcut in the loop below when a Py_None is encountered in the list
+  const QVariant nullVariant = QVariant( QVariant::Int );
```
Comment (troopa81) [inline python/PyQt6/core/auto_generated/qgsattributes.sip.in:77]:
```suggestion
  const QVariant nullVariant( QVariant::Int );
```

**Comment by nyalldawson** [comment]
@troopa81 good to merge now?

**Review by troopa81 — APPROVED** [review]

**Pull request merged**